lib/inventory/metadata/metadata.go

```diff
@@ -0,0 +1,334 @@
+	// It is configurable so that it can be mocked in tests.
+	httpDo func(req *http.Request) (*http.Response, error)
+	// kubeClient is a kubernetes client used to retrieve the
+	// server version.
+	// It is configurable so that it can be mocked in tests.
+	kubeClient kubernetes.Interface
+}
+
+// setDefaults sets the values of readFile and execCommand to the ones in the
+// standard library. Having these two methods configurable allows us to mock
+// them in tests.
+func (c *AgentMetadataFetchConfig) setDefaults() {
+	if c.Context == nil {
+		c.Context = context.Background()
+	}
+	if c.getenv == nil {
+		c.getenv = os.Getenv
+	}
+	if c.readFile == nil {
+		c.readFile = os.ReadFile
+	}
+	if c.execCommand == nil {
+		c.execCommand = func(name string, args ...string) ([]byte, error) {
+			return exec.Command(name, args...).Output()
+		}
+	}
+	if c.httpDo == nil {
+		c.httpDo = func(req *http.Request) (*http.Response, error) {
+			client, err := defaults.HTTPClient()
+			if err != nil {
+				return nil, err
+			}
+			client.Timeout = 5 * time.Second
+			return client.Do(req)
+		}
+	}
+	if c.kubeClient == nil {
+		c.kubeClient = getKubeClient()
+	}
+}
+
+// getKubeClient returns a kubernetes client in case the agent is running on
+// kubernetes. It returns nil otherwise.
+func getKubeClient() kubernetes.Interface {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		log.Debugf("Failed to get kubernetes cluster config: %s", err)
+		return nil
+	}
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Debugf("Failed to create kubernetes client: %s", err)
+		return nil
+	}
+	return client
+}
+
+// FetchAgentMetadata fetches and calculates all agent metadata we are interested
+// in tracking.
+func FetchAgentMetadata(c *AgentMetadataFetchConfig) proto.UpstreamInventoryAgentMetadata {
+	c.setDefaults()
+	return proto.UpstreamInventoryAgentMetadata{
+		OS:                    c.fetchOS(),
+		OSVersionInfo:         c.fetchOSVersionInfo(),
+		HostArchitectureInfo:  c.fetchHostArchitectureInfo(),
+		GlibcVersionInfo:      c.fetchGlibcVersionInfo(),
+		InstallMethods:        c.fetchInstallMethods(),
+		ContainerRuntime:      c.fetchContainerRuntime(),
+		ContainerOrchestrator: c.fetchContainerOrchestrator(),
+		CloudEnvironment:      c.fetchCloudEnvironment(),
+	}
+}
+
+// FetchServices computes the list of access protocols enabled at the agent from
+// the list of system roles present in the hello message.
+func FetchServices(systemRoles []types.SystemRole) []string {
+	var services []string
+	for _, svc := range systemRoles {
+		switch svc {
+		case types.RoleNode:
+			services = append(services, nodeService)
+		case types.RoleKube:
+			services = append(services, kubeService)
+		case types.RoleApp:
+			services = append(services, appService)
+		case types.RoleDatabase:
+			services = append(services, dbService)
+		case types.RoleWindowsDesktop:
+			services = append(services, windowsDesktopService)
+		}
+	}
+	return services
+}
```

Comment:
Nit: Probably better to avoid needing to update `inventory/metadata` any time we add a new system role.  I'd suggest adding a method like `IsAccessService` to `types.SystemRole`.  There is existing precedence for using the methods of `types.SystemRole` to distinguish what kind of role it is (e.g. `role.IsServerRole()`, which is like what you are doing here, but also includes auth/proxy).

Similarly, I'd either use the canonical names, or apply a standardized transform (e.g. `strings.ToLower(...)`), rather than hard-coding alternative names.

That way we can end up w/ something like this:

```golang
func NormalizeAndFilterServices(systemRoles []types.SystemdRole) []string {
    var services []string
    for _, svc := range systemRoles {
        if !svc.IsAccessService() {
            // skip service roles not directly associated with an access protocol (e.g. auth/proxy)
            continue
        }
        services = append(services, strings.ToLower(string(svc)))
    }
    return services
}
```

Reply:
Thanks! There was no motivation (IIRC) in having *only* the system roles that correspond to an access service, and so 64ec45d7d21c2fae8ae05c24524586c1664bc0d8 changes this in order to send all system roles to PreHog.